### PR TITLE
Add whirl backend

### PR DESCRIPTION
**Pull request by samcoppini**
This adds a backend for generating [Whirl](https://bigzaphod.github.io/Whirl/) programs.

### CODE REVIEW

File: target/whirl.c [hunk 1]
```diff
@@ -0,0 +1,1221 @@
...
+
+static void generate_math_command(WhirlCodeSegment *segment, RingState *state, MathCmd cmd);
+
+// Generates the Whirl code to execute a command on the operations ring, taking
+// into account the current state of the command rings
+static void generate_op_command(WhirlCodeSegment *segment, RingState *state, OpCmd cmd) {
+    if (state->active_ring == MATH_RING) {
+        // If we're on the wrong ring, we need to execute a nop command to switch to
+        // the operations ring
+        generate_math_command(segment, state, OP_NOOP);
```
Comment (shinh) [inline target/whirl.c]: MATH_NOOP?

https://travis-ci.org/shinh/elvm/builds/645138426#L329

File: target/whirl.c [hunk 2]
```diff
@@ -0,0 +1,1221 @@
...
+    emit_zero(segment, state);
+}
+
+// Generates the Whirl code to execute a command on the math ring, taking into
+// account the current state of the command rings
+static void generate_math_command(WhirlCodeSegment *segment, RingState *state, MathCmd cmd) {
+    if (state->active_ring == OPERATION_RING) {
+        // If we're on the wrong ring, we need to execute a nop command to switch to
+        // the math ring
+        generate_op_command(segment, state, MATH_NOOP);
```
Comment (shinh) [inline target/whirl.c]: OP_NOOP?

https://travis-ci.org/shinh/elvm/builds/645138426#L333

**Comment by ComFreek** [comment]
@bigzaphod Seems like Whirl is your language. Soon, we can have a C compiler in Whirl! :smile:

File: target/whirl.c (hunk 1 above)
Comment (samcoppini) [reply target/whirl.c]: Whoops. I have fixed this now, along with the other instance where OP_NOOP and MATH_NOOP were switched.

**Comment by samcoppini** [comment]
I have added whirl to the list of backends, so this be ready to go

**Comment by shinh** [comment]
Yay, thanks!!

**Pull request merged**